cblox/include/cblox/core/submap_collection.h

```diff
@@ -0,0 +1,138 @@
+#define CBLOX_CORE_SUBMAP_COLLECTION_H_
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "./TsdfSubmapCollection.pb.h"
+#include "cblox/core/common.h"
+#include "cblox/core/tsdf_esdf_submap.h"
+
+namespace cblox {
+
+template <typename SubmapType>
+class SubmapCollection {
+ public:
+  typedef std::shared_ptr<SubmapCollection> Ptr;
+  typedef std::shared_ptr<const SubmapCollection> ConstPtr;
+
+  // Constructor. Constructs an empty submap collection map
+  explicit SubmapCollection(const typename SubmapType::Config &submap_config)
+      : submap_config_(submap_config) {}
+
+  // Constructor. Constructs a submap collection from a list of submaps
+  SubmapCollection(const typename SubmapType::Config &submap_config,
+                   const std::vector<typename SubmapType::Ptr> &tsdf_sub_maps);
+
+  // Gets a vector of the linked IDs
+  std::vector<SubmapID> getIDs() const;
+  bool exists(const SubmapID submap_id) const;
+
+  // Creates a new submap on the top of the collection
+  void createNewSubMap(const Transformation &T_M_S, const SubmapID submap_id);
+  SubmapID createNewSubMap(const Transformation &T_M_S);
+
+  // Create a new submap which duplicates an existing source submap
+  bool duplicateSubMap(const SubmapID source_submap_id,
+                       const SubmapID new_submap_id);
+
+  // Gets a const pointer to a raw submap
+  // NOTE(alexmillane): This function hard fails when the submap doesn't
+  // exist... This puts the onus on the caller to call exists() first. I don't
+  // like this but I can't see a solution.
+  const SubmapType &getSubMap(const SubmapID submap_id) const;
+  // Note(alexmillane): Unlike the above this function returns a nullptr when
+  // the map doesn't exist. No hard crash.
+  typename SubmapType::ConstPtr getSubMapConstPtrById(
+      const SubmapID submap_id) const;
+  // A list of the submaps
+  const std::vector<typename SubmapType::Ptr> getSubMaps() const;
+
+  // Interactions with the active submap
+  const SubmapType &getActiveSubMap() const;
+  const Transformation &getActiveSubMapPose() const;
+  const SubmapID getActiveSubMapID() const;
+
+  // Access the tsdf_map member of the active submap
+  TsdfMap::Ptr getActiveTsdfMapPtr();
+  const TsdfMap &getActiveTsdfMap() const;
+
+  // Interacting with the submap poses
+  bool setSubMapPose(const SubmapID submap_id, const Transformation &pose);
+  void setSubMapPoses(const TransformationVector &transforms);
+  bool getSubMapPose(const SubmapID submap_id, Transformation *pose_ptr) const;
+  void getSubMapPoses(TransformationVector* submap_poses) const;
+
+  // Generate the ESDF for the submap with the given ID. This method is only
+  // available if the SubmapType is equal to or derived from TsdfEsdfSubmap.
+  template <class ST = SubmapType>
+  typename std::enable_if<std::is_base_of<TsdfEsdfSubmap, ST>::value,
+                          bool>::type
+  generateEsdfById(const SubmapID submap_id) {
```

Comment:
My gut feeling is that this design is sub-optimal. It's just quite complicated and as I see it, not required. Can clients not simply ask for a submap, and then call generateEsdf() themselves, and that way we avoid all the template meta-programming? Am I missing something?

Reply:
If we add a getActiveSubMapPtr() method, clients will be able to call generateEsdf() which is non-const themselves.
The above code is a leftover from when I thought it'd be nice if SubmapCollection had sole control over its submaps and clients could only get read access through const refs.

Reply:
Thanks for this. I agree with your reasoning.